src/pixelino/pixelino/Tools/Bucket.swift

```diff
@@ -0,0 +1,131 @@
+//
+//  Bucket.swift
+//  pixelino
+//
+//  Created by Sandra & Johannes.
+//
+
+import Foundation
+import UIKit.UIGestureRecognizer
+
+/// This class resembles the typical bucket functionality of a drawing tool.
+/// Users can tap on a pixel, and the enclosing area of pixels that share the same base
+/// color as the pixel you tapped will be colored in a specified target color.
+class Bucket: Tool {
+    
+    /// Handles a tap which results in bucket-filling the area that encloses the pixel.
+    /// Look at `colorNeighboredPixels` for more information on the bucket area coloring
+    /// functionality.
+    func handleTapFrom(_ sender: UITapGestureRecognizer, _ controller: DrawingViewController) {
+        
+        if let canvasView = controller.canvasView {
+            
+            // Frequently used variables.
+            let canvasScene = canvasView.canvasScene
+            let canvas = canvasView.canvas
+            
+            // Calculate correct location in terms of canvas and corresponding pixels.
+            let touchLocation = sender.location(in: sender.view)
+            let touchLocationInScene = canvasView.convert(touchLocation, to: canvasScene)
+            
+            // Get the tapped pixel.
+            let nodes = canvasScene.nodes(at: touchLocationInScene)
+            
+            // Initialize a new GroupDrawCommand.
+            controller.groupDrawCommand = GroupDrawCommand()
+            
+            nodes.forEach({ (node) in
+                if let pixel = node as? Pixel {
+                    
+                    // Grab the pixel coordinates of the tapped pixel in order to set
+                    // a starting point for the bucket-fill coloring.
+                    let pixelCoordinates = canvas.getPosition(pixel: pixel)
+                    
+                    colorNeighboredPixels(canvas,
+                                          startCoordinate: pixelCoordinates,
+                                          baseColor: pixel.fillColor,
+                                          targetColor: controller.currentDrawingColor,
+                                          controller)
+                    
+                    controller.commandManager.execute(controller.groupDrawCommand)
+                }
+            })
+        }
+    }
+    
+    /// Recursive algorithm that traverses the canvas starting from a particular pixel based
```

Comment:
maybe hint that an iterative algorithm could handle larger images better (in the future)